package-system/OpenSSL/FindOpenSSL.cmake.template

```diff
@@ -6,29 +6,64 @@
 #
 #
 
-set(TARGET_WITH_NAMESPACE "3rdParty::OpenSSL")
-if (TARGET $${TARGET_WITH_NAMESPACE})
+set(OPENSSL_O3DE_NAMESPACE "3rdParty::OpenSSL")
+if (TARGET $${OPENSSL_O3DE_NAMESPACE})
     return()
 endif()
 
+set(SSL_TARGETNAME "OpenSSL::SSL")
+set(CRYPTO_TARGETNAME "OpenSSL::Crypto")
+
+# we're trying to be a drop-in replacement for the FindOpenSSL.cmake that is shipped
+# with CMake itself, so we set the same variables with the same uppercase for compatibility
+# for questions about these variables, see https://cmake.org/cmake/help/latest/module/FindOpenSSL.html
+set(OPENSSL_FOUND True)
 set(OPENSSL_INCLUDE_DIR $${CMAKE_CURRENT_LIST_DIR}/OpenSSL/include)
 set(OPENSSL_LIBS_DIR $${CMAKE_CURRENT_LIST_DIR}/OpenSSL/$$<$$<CONFIG:debug>:debug/>lib)
+set(OPENSSL_CRYPTO_LIBRARY $${OPENSSL_LIBS_DIR}/libcrypto$${CMAKE_STATIC_LIBRARY_SUFFIX})
+set(OPENSSL_CRYPTO_LIBRARIES
+    $${OPENSSL_CRYPTO_LIBRARY}
+    ${CRYPTO_LIBRARY_DEPENDENCIES})
+set(OPENSSL_SSL_LIBRARY $${OPENSSL_LIBS_DIR}/libssl$${CMAKE_STATIC_LIBRARY_SUFFIX})
+set(OPENSSL_SSL_LIBRARIES $${OPENSSL_SSL_LIBRARY})
+set(OPENSSL_LIBRARIES
+    $${OPENSSL_CRYPTO_LIBRARIES}
+    $${OPENSSL_SSL_LIBRARIES})
+set(OPENSSL_VERSION_STRING "1.1.1m")
+set(OPENSSL_VERSION $${OPENSSL_VERSION_STRING})
 
-set(OPENSSL_COMPILE_DEFINITIONS
-    OPENSSL_ENABLED
-${CUSTOM_ADDITIONAL_COMPILE_DEFINITIONS})
-
-set(OPENSSL_LIBRARY
-    "$${OPENSSL_LIBS_DIR}/libssl$${CMAKE_STATIC_LIBRARY_SUFFIX}"
-    "$${OPENSSL_LIBS_DIR}/libcrypto$${CMAKE_STATIC_LIBRARY_SUFFIX}"
-${CUSTOM_ADDITIONAL_LIBRARIES})
+add_library($${SSL_TARGETNAME} STATIC IMPORTED GLOBAL)
+set_target_properties($${SSL_TARGETNAME} PROPERTIES IMPORTED_LINK_INTERFACE_LANGUAGES "C")
+set_target_properties($${SSL_TARGETNAME} PROPERTIES IMPORTED_LOCATION "$${OPENSSL_CRYPTO_LIBRARY}")
 
-add_library($${TARGET_WITH_NAMESPACE} INTERFACE IMPORTED GLOBAL)
+add_library($${CRYPTO_TARGETNAME} STATIC IMPORTED GLOBAL)
+set_target_properties($${CRYPTO_TARGETNAME} PROPERTIES IMPORTED_LINK_INTERFACE_LANGUAGES "C")
+set_target_properties($${CRYPTO_TARGETNAME} PROPERTIES IMPORTED_LOCATION "$${OPENSSL_CRYPTO_LIBRARY}")
 
-ly_target_include_system_directories(TARGET $${TARGET_WITH_NAMESPACE} INTERFACE $${OPENSSL_INCLUDE_DIR})
+add_library($${OPENSSL_O3DE_NAMESPACE} INTERFACE IMPORTED GLOBAL)
+target_link_libraries($${OPENSSL_O3DE_NAMESPACE} INTERFACE $${OPENSSL_LIBRARIES})
```

Comment:
Its probably better to make the official targets the real ones ,and then create an _alias_ to define the open3dengine 3rdPArty:xxxx one.

To be more specific, according to the official FindOpenSSL.cmake from cmake, after this script runs we expect the following targets:
OpenSSL::SSL
OpenSSL::Crypto

Open3Dengine can add its own, ie,
3rdParty::OpenSSL or whatever it wants to call it.

I'd recommend making the normal cmake-expected stuff so that its compatibel with the intrerface in CMAke's official FindOpenSSL.cmake, and then after that, add aliases to those targets for O3DE usage, ie,
```
add_libary(${SSL_TARGETNAME} STATIC IMPORTED)
... define the SSL_TARGETNAME library
add_library$(CRYPTO_TARGETNAME} STATIC IMPORTED)
... define the CRYPTO library...

...
add_library(${OPENSSL_O3DE_NAMESPACE} ALIAS ${SSL_TARGETNAME})

Also do the crypto and SSL targets depend on each other?  If they do, decide which one depends on the aother and make sure it has a target_link_libraries.

Reply:
Sure, I'll check whether crypto and SSL targets depend on each other.

The reason I don't use an alias for OPENSSL_O3DE_NAMESPACE is that our target is the combination of SSL_TARGETNAME and CRYPTO_TARGETNAME.

Reply:
Then the best way to do it is probably something like
```
add_libary($${SSL_TARGETNAME} STATIC IMPORTED GLOBAL)
... define the SSL_TARGETNAME library
add_library($${CRYPTO_TARGETNAME} STATIC IMPORTED GLOBAL)
... define the CRYPTO library...
add_library($${OPENSSL_O3DE_NAMESPACE} INTERFACE IMPORTED GLOBAL)

# depend on the other 2 targets for the 3rdparty O3DE one.
target_link_libraries($${OPENSSL_O3DE_NAMESPACE} INTERFACE $${CRYPTO_TARGETNAME) $${SSL_TARGETNAME})
```
In other words, when working with _targets_ keep the dependencies to other targets where possible as opposed to raw libraries.